### PR TITLE
Forse ROS2 parameters to be static to meet RCLCPP API changes

### PR DESCRIPTION
**Pull request by AlexeyMerzlyakov**
This commit fixes build failures with latest ROS2 releases appeared due to changed and deprecated RCLCPP API.
In RCLCPP API [update](https://discourse.ros.org/t/feedback-required-should-parameters-be-able-to-change-of-type/18319/2) all ROS2 parameters are treated to be a static by-default. This leaded an impossibility of using the `declare_parameters("with_one_argument")` method without knowing a type of parameter. Additionally, these changes forced all static parameters by their design intentionally to be initialized with some value.

This commit adopts Nav2 stack parameters for the changes from above and fixes related tests.

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #2220 |
| Primary OS tested on | Ubuntu 20.04 |
| Robotic platform tested on | TB3 simulation in Gazebo |

---

## Description of contribution in a few bullet points

* Adopted `declare_parameter()` direct calls for static parameters.
* Splitted `declare_parameter_if_not_declared()` routine into two implementations: first will be working with definite input parameter values, second - with defined type of parameters.
* Adopted dependent `declareParameter()` and `get_plugin_type_param()` routines.
* Switched to exception handling in some places which have a logic of processing not-defined parameter values.
* Added routines description in `nav2_util/include/nav2_util/node_utils.hpp`.
* Fixes testcases.

### CODE REVIEW

**Comment by AlexeyMerzlyakov** [comment]
Build fails with following message:
```
In file included from /opt/overlay_ws/src/navigation2/nav2_util/src/node_utils.cpp:15:
/opt/overlay_ws/src/navigation2/nav2_util/include/nav2_util/node_utils.hpp: In function ‘std::string nav2_util::get_plugin_type_param(NodeT, const string&)’:
/opt/overlay_ws/src/navigation2/nav2_util/include/nav2_util/node_utils.hpp:145:32: error: ‘NoParameterOverrideProvided’ in namespace ‘rclcpp::exceptions’ does not name a type
  145 |   } catch (rclcpp::exceptions::NoParameterOverrideProvided & ex) {
      |                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~
```
It looks like latest ROS2 changes are not in the Circle CI containers yet.

**Comment by SteveMacenski** [comment]
https://hub.docker.com/repository/docker/rosplanning/navigation2/tags?page=1&ordering=last_updated they seem to be up to date to me, only 8 hours since the last build, but it looks like the `nightly` hasn't been building for 21 days https://hub.docker.com/r/osrf/ros2/tags?page=1&ordering=last_updated

**Comment by SteveMacenski** [comment]
Otherwise LGTM

File: nav2_costmap_2d/plugins/costmap_filters/costmap_filter.cpp [hunk 1]
```diff
@@ -62,17 +62,17 @@ void CostmapFilter::onInitialize()
...
   declareParameter("enabled", rclcpp::ParameterValue(true));
-  declareParameter("filter_info_topic");
-  declareParameter("transform_tolerance", rclcpp::ParameterValue(0.1));
-
-  // Get parameters
-  node->get_parameter(name_ + "." + "enabled", enabled_);
   try {
-    filter_info_topic_ = node->get_parameter(name_ + "." + "filter_info_topic").as_string();
-  } catch (rclcpp::exceptions::ParameterNotDeclaredException & ex) {
+    declareParameter("filter_info_topic", rclcpp::PARAMETER_STRING);
```
[the remark this answers is not in the document]
Comment (AlexeyMerzlyakov) [reply nav2_costmap_2d/plugins/costmap_filters/costmap_filter.cpp:66]: Done in https://github.com/ros-planning/navigation2/pull/2239/commits/39bbc0956be6456631d0451624cee2cec0751efc

**Comment by AlexeyMerzlyakov** [comment]
I saw https://github.com/osrf/docker_images/issues/512 has been resolved! Now Circle CI is using latest ros2 rolling release.

With latest ROS2 changes the build of nav2 dependent packages fails on `gazebo_ros` with following message:
```
--- stderr: gazebo_ros                                                      
CMake Error at /opt/ros/rolling/share/ament_cmake_python/cmake/ament_python_install_package.cmake:98 (add_custom_target):
  add_custom_target called with invalid target name
  "ament_cmake_python_symlink_scripts/".  Target names may not contain a
  slash.  Use ADD_CUSTOM_COMMAND to generate files.
Call Stack (most recent call first):
  /opt/ros/rolling/share/ament_cmake_python/cmake/ament_python_install_package.cmake:33 (_ament_cmake_python_install_package)
  CMakeLists.txt:39 (ament_python_install_package)
```

This is fixed if simply apply the following patch to https://github.com/ros-simulation/gazebo_ros_pkgs.git repository:
```
diff --git a/gazebo_ros/CMakeLists.txt b/gazebo_ros/CMakeLists.txt
index 361f6de..fb0d244 100644
--- a/gazebo_ros/CMakeLists.txt
+++ b/gazebo_ros/CMakeLists.txt
@@ -36,7 +36,7 @@ link_directories(
   ${gazebo_dev_LIBRARY_DIRS}
 )
 
-ament_python_install_package(scripts/)
+ament_python_install_package(scripts)
 
 # gazebo_ros_node
 add_library(gazebo_ros_node SHARED
```
After that build of nav2 dependent packages and nav2 passes (checked on Circle CI container, debug build).

**Comment by AlexeyMerzlyakov** [comment]
Update: the fix was applied in https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1251 at the same time while I writing this message. Re-triggering build.

**Review by SteveMacenski — APPROVED** [review]

**Comment by SteveMacenski** [comment]
Build will take a few more hours to come back up. The nightly triggers around 6pm (if I recall) my time and then building the containers themselves will take several hours. I'll retrigger in the morning and merge when build is complete!

**Pull request merged**